### PR TITLE
Add ability to filter attachments/links by content

### LINKED ISSUE #15: Attachment/URL filtering
Allow the addition/removal of SHA512 hashes to a content filter. If a person uploads an attachment or links to a file with a SHA512 hash matching any in the filter, that message is either flagged or deleted, similar to the text filter. Similarly, any messages posted by people with the "Administrator" permission are not to be considered.

### PR DESCRIPTION
**Pull request by ammongit**
Fixes #15 

Enables the "content filter", allowing filtering by SHA1 hash. This applies to both attachments and to any links specified in the message. Like the text filter, there are the flag, block, and jail levels, however all content filters are applied guild-wise. Filter immunity is identical to text filters.

### CODE REVIEW

File: futaba/cogs/filter/check.py [hunk 1]
```diff
@@ -87,18 +116,60 @@ def journal_violation(journal, message, filter_type, flagged_content):
...
+            hashsums[digest] = (binio, url)
+
+    for hashsum, filter_type in cog.content_filters[message.guild].items():
+        try:
+            binio, url = hashsums[hashsum]
+        except KeyError:
+            # Hash sum not found, not a match
+            continue
+
+        if triggered is None or filter_type.value > triggered.filter_type.value:
```
Comment (raylu) [inline futaba/cogs/filter/check.py]: if you're only going to give the first one anyway, might as well break in here

**Comment by ammongit** [comment]
Cryptographic security isn't necessary, we just want to generally avoid collisions (so we don't filter things that weren't intended to be blocked). I changed from MD5 to SHA1 since it's _slightly_ better at avoiding collisions for an insignificant increase in hash length.

Do CRCs have issues with hash collisions?

File: futaba/cogs/filter/check.py (hunk 1 above)
Comment (raylu) [reply futaba/cogs/filter/check.py]: oh whoops

**Comment by raylu** [comment]
CRCs are much worse than any of the above at hash collisions. in particular, it's very easy to generate a hash collision for any given CRC32. for this application, it just means it's very easy to get yourself dunced or whatever

the advantage is that CRC32's are way faster to calculate

**Comment by ammongit** [comment]
Well given our bottleneck is downloading the file rather than calculating the hash, I don't see performance as a reason to switch. Thanks for the suggestion though.

**Pull request merged**